src/deserialization.hpp

```diff
@@ -256,185 +355,101 @@ class ConcreteTensorProtoDeserializator {
+        SPDLOG_TRACE("Extracting output: {}", name);
+        return request.getOutput(name.c_str(), tensor);
+    }
+};
 
-template <class TensorProtoDeserializator>
-ov::Tensor deserializeTensorProto(
-    const ::KFSRequest::InferInputTensor& requestInput,
-    const std::shared_ptr<const TensorInfo>& tensorInfo,
-    const std::string* buffer) {
-    return TensorProtoDeserializator::deserializeTensorProto(requestInput, tensorInfo, buffer);
-}
+template <>
+class RequestTensorExtractor<InferenceRequest, const InferenceTensor**, ExtractChoice::EXTRACT_INPUT> {
+public:
+    static Status extract(const InferenceRequest& request, const std::string& name, const InferenceTensor** tensor) {
+        SPDLOG_TRACE("Extracting input: {}", name);
+        return request.getInput(name.c_str(), tensor);
+    }
+};
 
-template <class TensorProtoDeserializator>
-ov::Tensor deserializeTensorProto(
-    const InferenceTensor& requestInput,
-    const std::shared_ptr<const TensorInfo>& tensorInfo) {
-    return TensorProtoDeserializator::deserializeTensorProto(requestInput, tensorInfo);
+// due to header included in many places function below is not used in all cpp files ...
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
+template <>  // TODO separate for different choice
+std::tuple<ovms::Status, const typename RequestTraits<ovms::InferenceRequest>::TensorType*, const std::string*> getRequestTensorPtr(const ovms::InferenceRequest& request, const std::string& name, ExtractChoice extractChoice) {
+    const InferenceTensor* requestTensorPtr{nullptr};
+    ovms::Status status;
+    switch (extractChoice) {
+    case ExtractChoice::EXTRACT_INPUT: {
+        status = RequestTensorExtractor<InferenceRequest, const InferenceTensor**, ExtractChoice::EXTRACT_INPUT>::extract(request, name, &requestTensorPtr);
+        break;
+    }
+    case ExtractChoice::EXTRACT_OUTPUT: {
+        status = RequestTensorExtractor<InferenceRequest, const InferenceTensor**, ExtractChoice::EXTRACT_OUTPUT>::extract(request, name, &requestTensorPtr);
+        break;
+    }
+    }
+    if ((!status.ok() || requestTensorPtr == nullptr)) {
+        SPDLOG_DEBUG("Failed to deserialize request. Validation of request failed");
+        return std::make_tuple(Status(StatusCode::INTERNAL_ERROR, "Failed to deserialize request"), nullptr, nullptr);
+    }
+    return std::make_tuple(Status(StatusCode::OK), requestTensorPtr, nullptr);
 }
+#pragma GCC diagnostic pop
 
-template <class Requester>
-class InputSink {
-    Requester requester;
-
-public:
-    InputSink(Requester requester) :
-        requester(requester) {}
-    Status give(const std::string& name, ov::Tensor& tensor);
-};
+#define RETURN_IF_EMPTY_TENSOR()                                           \
+    do {                                                                   \
+        if (!tensor) {                                                     \
```

Comment:
Why do we need "do while"?

Reply:
https://groups.google.com/g/comp.lang.c/c/xGZxls194mI/m/dEIpTKz2okMJ?pli=1